torch/lib/THCUNN/SpatialDepthwiseConvolution.cu

```diff
@@ -30,30 +40,42 @@ __global__ void spatialDepthwiseConvolutionUpdateOutput(
 {
   const int channelStride = outputHeight * outputWidth;
   const int batchStride = outputChannels * channelStride;
+  const int KW_LIMIT = (kSize !=0) ? kSize : kernelWidth;
+  const int KH_LIMIT = (kSize !=0) ? kSize : kernelHeight;
+  
 
   for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
        linearIndex < totalElements;
        linearIndex += gridDim.x * blockDim.x) {
-
-    const int n = linearIndex / batchStride;
-    const int c = (linearIndex / channelStride) % outputChannels;
-    const int h = (linearIndex / outputWidth) % outputHeight;
-    const int w = linearIndex % outputWidth;
-
-    const int inputChannel = c / depthwiseMultiplier;
-    const int inputChannels = outputChannels / depthwiseMultiplier;
+    int indtmp1 = linearIndex/outputWidth;
+    const int w = linearIndex - indtmp1 * outputWidth;
+    int indtmp2 = indtmp1/outputHeight;
+    const int h = indtmp1 - indtmp2 * outputHeight;
+    indtmp1 = indtmp2; 
+    indtmp2 = indtmp1/outputChannels;
+    const int c = indtmp1 - indtmp2 * outputChannels;
+    const int n = indtmp2;
+
+    int inputChannel = c; 
+    int inputChannels = outputChannels; 
+    if (depthwiseMultiplier !=1) {
+      inputChannel /= depthwiseMultiplier;
+      inputChannels /= depthwiseMultiplier;
+    }
 
     int weightOffset = c * kernelHeight * kernelWidth;
 
     AccT value = biasEnabled ? ScalarConvert<T, AccT>::to(bias.data()[c]) : ScalarConvert<int, AccT>::to(0);
-    for (int kH = 0; kH < kernelHeight; ++kH) {
-      for (int kW = 0; kW < kernelWidth; ++kW) {
+    const IndexType offset0 = (n * inputChannels + inputChannel) * inputHeight * inputWidth;
+#pragma unroll
+    for (int kH = 0; kH < KH_LIMIT; ++kH) {
+#pragma unroll
```

Comment:
is `#pragma unroll` a no-op if a loop doesn't have constant bounds?

Reply:
Yes